### PR TITLE
Lets AsyncSignedRequest and SignedRequest implement CopyableBuilder

### PR DESCRIPTION
**Pull request by cenedhryn**
## Motivation and Context
In the new signing flow, the input to the signer is a SignRequest (AsyncSignRequest) and the output is a SignedRequest (AsyncSignedRequest). SignRequest implements CopyableBuilder which means it supports `toBuilder` and `copy`, but SignedRequest does not. For use cases where a SignedRequest needs modification, for instance to manipulate headers in the SdkHttpRequest within the SignedRequest, it should implement this interface

## Modifications
- Added changes to SignedRequest / AsyncSignedRequest to implement `CopyableBuilder`
- Removed the anonymous implementation of SignedRequest / AsyncSignedRequest in the signer of `DefaultNoAuthAuthSchema` and instead lets that (anonymous) signer use the default builders
- Added unit tests for SignedRequest / AsyncSignedRequest. 

## Testing
Unit tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/signer/BaseSignedRequest.java [hunk 1]
```diff
@@ -20,6 +20,8 @@
 import software.amazon.awssdk.annotations.SdkPublicApi;
 import software.amazon.awssdk.annotations.ThreadSafe;
 import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
```
Comment (gosar) [inline core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/signer/BaseSignedRequest.java:24]: unnecessary import

File: core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/internal/signer/DefaultBaseSignedRequest.java [hunk 2]
```diff
@@ -15,6 +15,7 @@
 
 package software.amazon.awssdk.http.auth.spi.internal.signer;
 
+import java.util.Objects;
```
Comment (gosar) [inline core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/internal/signer/DefaultBaseSignedRequest.java:18]: unnecessary import

**Review by gosar — APPROVED** [review]

File: core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/internal/signer/DefaultBaseSignedRequest.java (hunk 2 above)
Comment (cenedhryn) [reply core/http-auth-spi/src/main/java/software/amazon/awssdk/http/auth/spi/internal/signer/DefaultBaseSignedRequest.java:18]: yeah I missed running checkstyle locally, fixing...

File: build-tools/src/main/java/software/amazon/awssdk/buildtools/findbugs/ToBuilderIsCorrect.java [hunk 3]
```diff
@@ -430,7 +430,11 @@ private void validateBuilderViaMethods(String buildableClassName, JavaClass buil
             if (builders.containsKey(concreteClass)) {
                 // We're invoking these methods on a known builder. Assume the method name matches the field name and validate
                 // based on that.
-                Set<String> builderFieldsForBuilder = new HashSet<>(builderFields.get(concreteClass));
+                List<String> concreteClassBuilderFields = builderFields.get(concreteClass);
+                if (concreteClassBuilderFields == null) {
+                    concreteClassBuilderFields = new ArrayList<>();
+                }
```
Comment (millems) [inline build-tools/src/main/java/software/amazon/awssdk/buildtools/findbugs/ToBuilderIsCorrect.java:436]: getOrDefault?

**Review by millems — APPROVED** [review]

File: build-tools/src/main/java/software/amazon/awssdk/buildtools/findbugs/ToBuilderIsCorrect.java (hunk 3 above)
Comment (cenedhryn) [reply build-tools/src/main/java/software/amazon/awssdk/buildtools/findbugs/ToBuilderIsCorrect.java:436]: That'd be nice(r)

**Pull request merged**